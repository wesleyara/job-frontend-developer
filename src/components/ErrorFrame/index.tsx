import { MdErrorOutline } from "react-icons/md";

interface ErrorFrameProps {
  error: Error | null;
}

export const ErrorFrame = ({ error }: ErrorFrameProps) => {
  return (
    <section
      data-cy="error-frame"
      className="mt-5 w-full px-3 flex flex-col items-center gap-3"
    >
      <MdErrorOutline size={50} />
      <span className="text-center">{error?.message}</span>
    </section>
  );
};
