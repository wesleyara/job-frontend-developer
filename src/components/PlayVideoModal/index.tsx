import { Dialog, DialogPanel, DialogTitle } from "@headlessui/react";
import { YoutubeDetailsType } from "~/types";
import { Dispatch, SetStateAction } from "react";

interface PlayVideoModalProps {
  item: YoutubeDetailsType;
  isOpen: boolean;
  setIsOpen: Dispatch<SetStateAction<boolean>>;
}

export const PlayVideoModal = ({
  item,
  isOpen,
  setIsOpen,
}: PlayVideoModalProps) => {
  return (
    <Dialog
      open={isOpen}
      onClose={() => setIsOpen(false)}
      className="relative z-50"
    >
      <div className="fixed inset-0 flex mx-auto items-center justify-center p-4">
        <DialogPanel className="space-y-4 border border-gray-300 bg-white p-12 max-w-[768px] w-[90%] text-center rounded-md relative">
          <button
            data-cy="close-youtube-modal"
            className="absolute top-2 right-3"
            onClick={() => setIsOpen(false)}
          >
            X
          </button>
          <DialogTitle className="text-lg">{item.snippet.title}</DialogTitle>

          <iframe
            data-cy="youtube-player"
            src={`https://www.youtube.com/embed/${item.id}`}
            title={item.snippet.title}
            className="inset-0 w-full h-[200px] sm:h-[400px]"
            allowFullScreen
          />
        </DialogPanel>
      </div>
    </Dialog>
  );
};
