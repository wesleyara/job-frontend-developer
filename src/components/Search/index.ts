import { SearchButton } from "./SearchButton";
import { SearchInput } from "./SearchInput";
import { SearchRoot } from "./SearchRoot";

export const Search = {
  Root: SearchRoot,
  Button: SearchButton,
  Input: SearchInput,
};
