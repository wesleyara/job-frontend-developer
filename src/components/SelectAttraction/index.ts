import { SelectAttractionButton } from "./SelectAttractionButton";
import { SelectAttractionContent } from "./SelectAttractionContent";
import { SelectAttractionRoot } from "./SelectAttractionRoot";

export const SelectAttraction = {
  Root: SelectAttractionRoot,
  Content: SelectAttractionContent,
  Button: SelectAttractionButton,
};
