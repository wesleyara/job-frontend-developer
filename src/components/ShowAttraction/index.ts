import { ShowAttractionBadges } from "./ShowAttractionBadges";
import { ShowAttractionInfo } from "./ShowAttractionInfo";
import { ShowAttractionRoot } from "./ShowAttractionRoot";

export const ShowAttraction = {
  Root: ShowAttractionRoot,
  Info: ShowAttractionInfo,
  Badges: ShowAttractionBadges,
};
