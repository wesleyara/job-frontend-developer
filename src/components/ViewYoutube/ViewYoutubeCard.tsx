"use client";

import { YoutubeDetailsType } from "~/types";
import Image from "next/image";
import { useState } from "react";
import { AiFillLike } from "react-icons/ai";
import { FaPlay, FaRegComment } from "react-icons/fa";
import { GrView } from "react-icons/gr";

interface ViewYoutubeCardProps {
  item: YoutubeDetailsType;
}

export const ViewYoutubeCard = ({ item }: ViewYoutubeCardProps) => {
  const [playHover, setPlayHover] = useState<boolean>(false);
  const [currentVideo, setCurrentVideo] = useState<string>("");

  const handleMouseEnter = (id: string) => {
    setCurrentVideo(id);
    setPlayHover(true);
  };

  const handleMouseLeave = () => {
    setCurrentVideo("");
    setPlayHover(false);
  };

  const showPlay = (id: string) => {
    return playHover && currentVideo === id;
  };

  const renderCounters = (count: string) => {
    const countNumber = Number(count);

    if (countNumber > 1000000000) {
      return `${(countNumber / 1000000000).toFixed(1)} bi`;
    }

    if (countNumber > 1000000) {
      return `${(countNumber / 1000000).toFixed(1)} mi`;
    }

    if (countNumber > 1000) {
      return `${(countNumber / 1000).toFixed(1)} mil`;
    }

    return countNumber;
  };

  return (
    <article
      key={item.id}
      className="flex flex-col gap-3 p-3 rounded-md border border-gray-300 bg-white w-full items-center"
    >
      <span
        className="relative cursor-pointer"
        onMouseEnter={() => handleMouseEnter(item.id)}
        onMouseLeave={() => handleMouseLeave()}
      >
        <FaPlay
          className={`play-tooltip ${showPlay(item.id) ? "show" : ""}`}
          size={40}
          color="#fff"
        />
        <Image
          src={item.snippet.thumbnails.medium.url}
          alt={item.snippet.title}
          width={320}
          height={180}
        />
      </span>
      <h3 className="text-lg font-bold">{item.snippet.title}</h3>
      <span className="flex gap-5">
        <span className="flex gap-1 items-center justify-center">
          <GrView /> {renderCounters(item.statistics.viewCount)}
        </span>
        <span className="flex gap-1 items-center justify-center">
          <AiFillLike /> {renderCounters(item.statistics.likeCount)}
        </span>
        <span className="flex gap-1 items-center justify-center">
          <FaRegComment /> {renderCounters(item.statistics.commentCount)}
        </span>
      </span>
      {item.snippet.description && (
        <span className="text-center">
          {item.snippet.description.slice(0, 100)}...
        </span>
      )}
    </article>
  );
};
