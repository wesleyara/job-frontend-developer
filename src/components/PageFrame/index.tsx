"use client";

import { useData } from "~/context/DataContext";
import { useState } from "react";
import { FaSearch } from "react-icons/fa";

import { ErrorFrame } from "../ErrorFrame";
import { Search } from "../Search";
import { SelectAttraction } from "../SelectAttraction";
import { ShowAttraction } from "../ShowAttraction";
import { ViewYoutube } from "../ViewYoutube";

export const PageFrame = () => {
  const {
    searchQuery,
    dataTicketSearch,
    dataYoutubeVideosDetails,
    selectedAttraction,
    isLoadingTicketSearch,
    isLoadingYoutubeVideos,
    isLoadingYoutubeVideosDetails,
    errorTicketSearch,
    errorYoutubeVideos,
    errorYoutubeVideosDetails,
    handleChange,
    handleSelectAttraction,
  } = useData();

  const [ratioSelected, setRatioSelected] = useState<string>("");
  const [seeMore, setSeeMore] = useState<boolean>(false);

  const enableToRender = dataTicketSearch && dataTicketSearch.length > 0;

  const render = () => {
    if (isLoadingTicketSearch) {
      return <p className="mt-10">Carregando...</p>;
    }

    if (errorTicketSearch || errorYoutubeVideos || errorYoutubeVideosDetails) {
      return <ErrorFrame error={errorTicketSearch} />;
    }

    if (selectedAttraction) {
      const imagesPerRatio: any = [];

      selectedAttraction.images.forEach(image => {
        const ratio = image.ratio;
        const find = imagesPerRatio.find((item: any) => item.ratio === ratio);

        if (find) {
          find.images.push(image);
        } else {
          imagesPerRatio.push({
            ratio,
            images: [image],
          });
        }
      });

      const selected = imagesPerRatio.find(
        (item: any) => item.ratio === ratioSelected,
      );

      return (
        <>
          <ShowAttraction.Root onClick={() => setSeeMore(prev => !prev)}>
            {!seeMore ? (
              <>
                <ShowAttraction.Info selectedAttraction={selectedAttraction} />
                <ShowAttraction.Badges
                  classifications={selectedAttraction.classifications}
                />
              </>
            ) : (
              <ShowAttraction.Carousel
                setRatioSelected={setRatioSelected}
                selected={selected}
                imagesPerRatio={imagesPerRatio}
                ratioSelected={ratioSelected}
              />
            )}
          </ShowAttraction.Root>

          <span className="flex mt-5">
            {isLoadingYoutubeVideos || isLoadingYoutubeVideosDetails ? (
              <p>Carregando...</p>
            ) : (
              !!dataYoutubeVideosDetails && (
                <ViewYoutube.Root>
                  {dataYoutubeVideosDetails?.map(item => (
                    <ViewYoutube.Card key={item.id} item={item} />
                  ))}
                </ViewYoutube.Root>
              )
            )}
          </span>
        </>
      );
    }

    if (enableToRender) {
      return (
        <div className="flex flex-col gap-3 mt-5 overflow-auto md:w-[600px] w-[90%]">
          {dataTicketSearch.map(item => (
            <SelectAttraction.Root key={item.id}>
              <SelectAttraction.Content item={item} />
              <SelectAttraction.Button
                onClick={() => handleSelectAttraction(item)}
              />
            </SelectAttraction.Root>
          ))}
        </div>
      );
    }
  };

  return (
    <main className="flex flex-col items-center my-24">
      <Search.Root>
        <Search.Input onChange={handleChange} value={searchQuery} />
        <Search.Button icon={FaSearch} />
      </Search.Root>

      {render()}
    </main>
  );
};
