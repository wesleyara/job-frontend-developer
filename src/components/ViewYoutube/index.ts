import { ViewYoutubeCard } from "./ViewYoutubeCard";
import { ViewYoutubeRoot } from "./ViewYoutubeRoot";

export const ViewYoutube = {
  Root: ViewYoutubeRoot,
  Card: ViewYoutubeCard,
};
