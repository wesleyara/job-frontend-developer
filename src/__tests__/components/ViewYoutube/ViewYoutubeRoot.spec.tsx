import { render } from "@testing-library/react";
import { ViewYoutube } from "~/components/ViewYoutube";
import { beforeEach, describe, expect, it } from "vitest";

describe("ViewYoutubeRoot unit test", () => {
  let sut: JSX.Element;

  beforeEach(() => {
    sut = (
      <ViewYoutube.Root>
        <span />
      </ViewYoutube.Root>
    );
  });

  it("should be defined", () => {
    expect(sut).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(sut);

    expect(container).toMatchSnapshot();
  });

  it("should have a section element", () => {
    const { container } = render(sut);

    expect(container.querySelector("section")).not.toBeNull();
  });

  it("should have a child element", () => {
    const { container } = render(sut);

    expect(container.querySelector("span")).not.toBeNull();
  });

  it("should have a tailwind class name", () => {
    const { container } = render(sut);

    const element = container.querySelector(".flex");

    expect(element).not.toBeNull();
  });
});
