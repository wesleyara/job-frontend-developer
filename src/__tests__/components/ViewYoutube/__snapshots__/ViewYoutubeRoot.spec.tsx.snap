// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ViewYoutubeRoot unit test > should render correctly 1`] = `
<div>
  <section
    class="flex flex-col gap-3 w-[90%] mx-auto max-w-[400px] items-center"
  >
    <span />
  </section>
</div>
`;
